cmd/daemonset-check/run_check.go

```diff
@@ -0,0 +1,1008 @@
+package main
+
+import (
+	"context"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	appsv1 "k8s.io/api/apps/v1"
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/Comcast/kuberhealthy/v2/pkg/checks/external/util"
+)
+
+// Globals for revealing daemonsets that fail to be removed or
+// nodes with daemonset pods that fail to come up or be removed
+var daemonsetList []appsv1.DaemonSet
+var nodesMissingDSPod []string
+var podRemovalList *apiv1.PodList
+
+// runCheck runs pre-check cleanup and then the full daemonset check. If successful, reports OK to kuberhealthy.
+func runCheck() {
+
+	log.Infoln("Running pre-check cleanup. Deleting any rogue daemonset or daemonset pods before running daemonSet check.")
+	err := cleanUp()
+	if err != nil {
+		log.Errorln("Error running pre-check cleanup:", err)
+		os.Exit(1)
+	}
+
+	log.Infoln("Running daemonset check")
+
+	err = runDaemonsetCheck()
+	if err != nil {
+		// If shutdown signal is received, proceed to shutdown instead of exiting.
+		if strings.Contains(err.Error(), "Failed to complete check due to an interrupt signal") {
+			return
+		}
+		log.Errorln("Error running check:", err)
+		os.Exit(1)
+	}
+
+	log.Infoln("Finished running checks, reporting OK to Kuberhealthy ")
+	reportOKToKuberhealthy()
+}
+
+// cleanUp triggers check clean up and waits for all rogue daemonsets to clear
+func cleanUp() error {
+
+	log.Debugln("Allowing this clean up", checkTimeLimit, "to finish.")
+	cleanUpCtx, cleanUpCtxCancel = context.WithTimeout(context.Background(), checkTimeLimit)
+
+	err := runCheckCleanUp()
+	if err != nil {
+		errorMessage := "Error running cleanup: " + err.Error()
+		log.Errorln(errorMessage)
+		reportErrorsToKuberhealthy([]string{"kuberhealthy/daemonset: " + errorMessage})
+		return errors.New(errorMessage)
+	}
+
+	// waiting for all daemonSets to be gone...
+	log.Infoln("Waiting for all daemonSets or daemonset pods to clean up")
+
+	outChan := make(chan error, 10)
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		log.Debugln("Worker: waitForAllDaemonsetsToClear started")
+		defer wg.Done()
+		defer close(outChan)
+		outChan <- waitForAllDaemonsetsToClear()
+		wg.Wait()
```

Comment:
I think this `wg.Wait()` will always block here since you `defer`d the `wg.Done()` within the same function. `wg.Wait()` is going to block until all `wg.Done()`s are called, but since you defer it here, it would always be called after the `wg.Wait()`